src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,163 @@
+                            binaryExpression.Right is MemberAccessExpressionSyntax)
+                        {
+                            memberAccess = (MemberAccessExpressionSyntax)binaryExpression.Right;
+
+                            var incrementor = forStatement.Incrementors[0];
+                            return TryGetStepValue(
+                                iterationVariable, incrementor, out stepValueExpressionOpt, cancellationToken);
+                        }
+                    }
+                }
+            }
+
+            iterationVariable = default;
+            memberAccess = default;
+            initializer = default;
+            stepValueExpressionOpt = default;
+            return false;
+        }
+
+        private static bool TryGetStepValue(
+            SyntaxToken iterationVariable, ExpressionSyntax incrementor,
+            out ExpressionSyntax stepValue, CancellationToken cancellationToken)
+        {
+            // support
+            //  x++
+            //  ++x
+            //  x += constant_1
+
+            ExpressionSyntax operand;
+            switch (incrementor.Kind())
+            {
+            case SyntaxKind.PostIncrementExpression:
+                operand = ((PostfixUnaryExpressionSyntax)incrementor).Operand;
+                stepValue = default;
+                break;
+
+            case SyntaxKind.PreIncrementExpression:
+                operand = ((PrefixUnaryExpressionSyntax)incrementor).Operand;
+                stepValue = default;
+                break;
+
+            case SyntaxKind.AddAssignmentExpression:
+                var assignment = (AssignmentExpressionSyntax)incrementor;
+                operand = assignment.Left;
+                stepValue = assignment.Right;
+                break;
+
+            default:
+                stepValue = null;
+                return false;
+            }
+
+            return operand is IdentifierNameSyntax identifierName &&
+                identifierName.Identifier.ValueText == iterationVariable.ValueText;
+        }
+
+        protected override SyntaxNode ConvertForNode(
+            ForStatementSyntax forStatement, TypeSyntax typeNode, 
+            SyntaxToken foreachIdentifier, ExpressionSyntax collectionExpression,
+            ITypeSymbol iterationVariableType, OptionSet optionSet)
+        {
+            if (typeNode == null)
+            {
+                // types are not apparent in foreach statements.
+                var isBuiltInTypeContext = TypeStyleHelper.IsBuiltInType(iterationVariableType);
+                if (TypeStyleHelper.IsImplicitStylePreferred(
+                        optionSet, isBuiltInTypeContext, isTypeApparentContext: false))
+                {
+                    typeNode = SyntaxFactory.IdentifierName("var");
+                }
+                else
+                {
+                    typeNode = (TypeSyntax)CSharpSyntaxGenerator.Instance.TypeExpression(iterationVariableType);
```

Comment:
is Generator has some state in it? why TypeExpression is not static method? anyway, good to know we have this.

Reply:
> is Generator has some state in it? why TypeExpression is not static method?

Because Generator is an abstract type that is shared across VB and C#.  As such, all members on it are instance so that each language can implement accordingly.  That's also why it just returns SyntaxNode.